### PR TITLE
PR: LiveView + Alpine.js select input

### PR DESCRIPTION
**Pull request by SimonLab**
Create basic structure for input select
ref: #6

### CODE REVIEW

**Comment by SimonLab** [comment]
PR ready for review, documentation at: https://github.com/dwyl/learn-alpine.js/blob/select-input/select-input.md

**Comment by nelsonic** [comment]
```sh
git checkout select-input
mix deps.get
mix ecto.reset
mix c
```

Output:

```sh
...........

  1) test Index saves new item (AppWeb.ItemLiveTest)
     test/app_web/live/item_live_test.exs:24
     Assertion with =~ failed
     code:  assert html =~ "Item created successfully"
     left:  "<html lang=\"en\"><head><meta charset=\"utf-8\"/><meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\"/><meta name=\"csrf-token\" content=\"ak0hegARZnRpFhYJPyhCHEApDm4IVVgbY9LKjy_7YBNNPz1d0SWCz4-W\"/><script defer=\"defer\" src=\"https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js\">\n    </script><script src=\"https://unpkg.com/@alpinejs/mask@3.x.x/dist/cdn.min.js\">\n    </script><title data-suffix=\" · Phoenix Framework\">Listing Items · Phoenix Framework</title><link phx-track-static=\"phx-track-static\" rel=\"stylesheet\" href=\"/assets/app.css\"/><script defer=\"defer\" phx-track-static=\"phx-track-static\" type=\"text/javascript\" src=\"/assets/app.js\">\n    </script></head><body><nav class=\"container mx-auto my-4\"><a class=\"font-bold text-slate-500 text-lg\" href=\"/\">Alpines.js Examples</a></nav><div data-phx-main=\"true\" data-phx-session=\"SFMyNTY.g2gDaAJhBXQAAAAIZAACaWRtAAAAFHBoeC1GeWJnOWswZXF2OEItd0hCZAAMbGl2ZV9zZXNzaW9uaAJkAAdkZWZhdWx0bggAlnRBwBTgJhdkAApwYXJlbnRfcGlkZAADbmlsZAAIcm9vdF9waWRkAANuaWxkAAlyb290X3ZpZXdkABxFbGl4aXIuQXBwV2ViLkl0ZW1MaXZlLkluZGV4ZAAGcm91dGVyZAAURWxpeGlyLkFwcFdlYi5Sb3V0ZXJkAAdzZXNzaW9udAAAAABkAAR2aWV3ZAAcRWxpeGlyLkFwcFdlYi5JdGVtTGl2ZS5JbmRleG4GACBvhmyEAWIAAVGA.Bv0prOpP3KTdIBWV7WqGtMo3v7lnKpKAtM0XK20H6Ko\" data-phx-static=\"SFMyNTY.g2gDaAJhBXQAAAADZAAKYXNzaWduX25ld2pkAAVmbGFzaHQAAAABbQAAAARpbmZvbQAAABlJdGVtIGNyZWF0ZWQgc3VjY2Vzc2Z1bGx5ZAACaWRtAAAAFHBoeC1GeWJnOWswZXF2OEItd0hCbgYAIG-GbIQBYgABUYA.hYc05Zms_2anZpUfjNkv-r8WdRqx-hsGa2LuXzCP-98\" id=\"phx-Fybg9k0eqv8B-wHB\"><main class=\"container mx-auto\"><h1 class=\"font-bold my-3\">Listing Items</h1><table class=\"min-w-full overflow-hidden divide-y ring-1 ring-gray-200 dark:ring-0 divide-gray-200 rounded-sm table-auto dark:divide-y-0 dark:divide-gray-800 sm:rounded\"><thead><tr><th class=\"px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-300\">Text</th><th class=\"px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-300\">Index</th></tr></thead><tbody id=\"items\" phx-hook=\"Items\" x-data=\"{selectedItem: null}\"><tr id=\"item-9\" data-id=\"9\" class=\"item border-b dark:border-gray-700 bg-white dark:bg-gray-800 last:border-none\" draggable=\"true\" x-data=\"{selected: false}\" x-on:dragstart=\"selected = true; $dispatch(&#39;highlight&#39;, {id: $el.id}); selectedItem = $el\" x-on:dragend=\"selected = false; $dispatch(&#39;remove-highlight&#39;, {id: $el.id}); selectedItem = null; $dispatch(&#39;update-indexes&#39;)\" x-bind:class=\"selected ? &#39;cursor-grabbing&#39; : &#39;cursor-grab&#39;\" x-on:dragover.throttle=\"$dispatch(&#39;dragoverItem&#39;, {selectedItemId: selectedItem.id, currentItemId: $el.id})\" data-highlight=\"[[&quot;add_class&quot;,{&quot;names&quot;:[&quot;bg-yellow-300&quot;],&quot;time&quot;:200,&quot;to&quot;:null,&quot;transition&quot;:[[],[],[]]}]]\" data-remove-highlight=\"[[&quot;remove_class&quot;,{&quot;names&quot;:[&quot;bg-yellow-300&quot;],&quot;time&quot;:200,&quot;to&quot;:null,&quot;transition&quot;:[[],[],[]]}]]\"><td class=\"px-6 py-4 text-sm text-gray-500 dark:text-gray-400\">some text</td><td class=\"px-6 py-4 text-sm text-gray-500 dark:text-gray-400\">1</td></tr><tr id=\"item-10\" data-id=\"10\" class=\"item border-b dark:border-gray-700 bg-white dark:bg-gray-800 last:border-none\" draggable=\"true\" x-data=\"{selected: false}\" x-on:dragstart=\"selected = true; $dispatch(&#39;highlight&#39;, {id: $el.id}); selectedItem = $el\" x-on:dragend=\"selected = false; $dispatch(&#39;remove-highlight&#39;, {id: $el.id}); selectedItem = null; $dispatch(&#39;update-indexes&#39;)\" x-bind:class=\"selected ? &#39;cursor-grabbing&#39; : &#39;cursor-grab&#39;\" x-on:dragover.throttle=\"$dispatch(&#39;dragoverItem&#39;, {selectedItemId: selectedItem.id, currentItemId: $el.id})\" data-highlight=\"[[&quot;add_class&quot;,{&quot;names&quot;:[&quot;bg-yellow-300&quot;],&quot;time&quot;:200,&quot;to&quot;:null,&quot;transition&quot;:[[],[],[]]}]]\" data-remove-highlight=\"[[&quot;remove_class&quot;,{&quot;names&quot;:[&quot;bg-yellow-300&quot;],&quot;time&quot;:200,&quot;to&quot;:null,&quot;transition&quot;:[[],[],[]]}]]\"><td class=\"px-6 py-4 text-sm text-gray-500 dark:text-gray-400\">some text</td><td class=\"px-6 py-4 text-sm text-gray-500 dark:text-gray-400\">2</td></tr></tbody></table><span class=\"block w-32 text-center bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded my-3\"><a data-phx-link=\"patch\" data-phx-link-state=\"push\" href=\"/items/new\">New Item</a></span></main></div></body></html>"
     right: "Item created successfully"
     stacktrace:
       test/app_web/live/item_live_test.exs:38: (test)

.
Finished in 0.2 seconds (0.05s async, 0.1s sync)
13 tests, 1 failure

Randomized with seed 215559
----------------
COV    FILE                                        LINES RELEVANT   MISSED
  0.0% lib/app.ex                                      9        0        0
 75.0% lib/app/application.ex                         36        4        1
  0.0% lib/app/repo.ex                                 5        0        0
 52.4% lib/app/tasks.ex                              159       21       10
100.0% lib/app/tasks/item.ex                          18        2        0
  0.0% lib/app/tasks/person.ex                        19        2        2
  0.0% lib/app_web.ex                                110        1        1
 25.0% lib/app_web/controllers/page_controller.       19        4        3
  0.0% lib/app_web/endpoint.ex                        46        0        0
  0.0% lib/app_web/live/counter_live/index.ex         21        3        3
 41.7% lib/app_web/live/item_live/form_componen       55       12        7
 33.3% lib/app_web/live/item_live/index.ex           107       18       12
  0.0% lib/app_web/live/item_live/show.ex             21        5        5
 90.0% lib/app_web/live/live_helpers.ex               60       10        1
  0.0% lib/app_web/live/select_input_live/index       28        6        6
 30.8% lib/app_web/router.ex                          39       13        9
 80.0% lib/app_web/telemetry.ex                       71        5        1
 25.0% lib/app_web/views/error_helpers.ex             30        4        3
100.0% lib/app_web/views/error_view.ex                16        1        0
  0.0% lib/app_web/views/layout_view.ex                7        0        0
  0.0% lib/app_web/views/page_view.ex                  3        0        0
100.0% test/support/conn_case.ex                      38        2        0
 57.1% test/support/data_case.ex                      58        7        3
100.0% test/support/fixtures/tasks_fixtures.ex        20        2        0
[TOTAL]  45.1%
----------------
```

The important 3 files from the perspective of this PR are:

```sh
  0.0% lib/app/tasks/person.ex                        19        2        2
  0.0% lib/app_web/live/select_input_live/index       28        6        6
 30.8% lib/app_web/router.ex                          39       13        9
```

the first two were added by this PR and have **`0% coverage`** ... 0️⃣ 😢 
as much as I want to merge this PR because the UI and docs are good. 👌 
I feel like we should be making more of an effort to test things. 💭

File: select-input.md [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+</div>
+
+
+## Create People
+
+Let's create a `people` table:
+
+```sh
+mix ecto.gen.migration add_people_table
```
Comment (nelsonic) [inline select-input.md:15]: Any reason to not use [`mix phx.gen.schema`](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Schema.html) for this?

**Comment by nelsonic** [comment]
Deployed to Fly.io as per #14 just to confirm that it wasn't just "broken" on my `localhost` ...
https://alpine-demo.fly.dev/select-input
image

Keen to get this tidy. 🙏

**Comment by nelsonic** [comment]
When I attempt to type some text this wierd grey box covers the text:

image

Can demo for you on standup on Monday if you are unable to replicate it. 👌

**Review by nelsonic — DISMISSED** [review]
@SimonLab overall I'm pretty stoked with this. 🙌 
Would prefer if there were some basic tests. 💭 
But if we (you) open an issue to add them later, we can merge this. 👌

**Review by Arhell — APPROVED** [review]

**Comment by nelsonic** [comment]
@SimonLab https://community.fly.io/t/how-to-run-seeds-exs-for-elixir-phoenix-app/3591

File: select-input.md (hunk 1 above)
Comment (SimonLab) [reply select-input.md:15]: Yes I could have use this command, I think I forgot it exists

**Comment by SimonLab** [comment]
related to https://github.com/dwyl/learn-alpine.js/pull/18#issuecomment-1312526321

I've also spotted this, I thought it was specific to my Firefox.
It's difficutl to debug as it shows sporadically on my side and I can't inspect it to see what dom element it is.
It might be the browser autocomplete suggestions, I'll turn this off to see if it fixes this isssue, see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete

File: lib/app_web/router.ex [hunk 2]
```diff
@@ -19,7 +19,8 @@ defmodule AppWeb.Router do
 
     live "/items", ItemLive.Index, :index
     live "/items/new", ItemLive.Index, :new
-
+    
+    # routes created by Phoenix. Not used at the moment
```
Comment (nelsonic) [inline lib/app_web/router.ex:23]: maybe just comment out the unused routes? 💭

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  build:
+    name: Build and test
+    environment: dwylauth
```
Comment (nelsonic) [inline .github/workflows/ci.yml:12]: Removing this line as it's a hangover from copying this file from the `auth` project. 👌 
```suggestion
```

**Review by nelsonic — APPROVED** [review]
@SimonLab looks great! 😍 
Thanks for adding tests. ✅ 
You're becoming quite the `Alpine.js` expert. 😜

**Comment by SimonLab** [comment]
Thanks for the review, once deployed I will run the seeds on Fly.io to make sure we have some data for the select-input :+1:

**Comment by nelsonic** [comment]
Great PR. Thanks again! 🎉 
Looking forward to seeing the multi-select working on the fly demo! https://alpine-demo.fly.dev/select-input 🤞

**Comment by SimonLab** [comment]
```sh
flyctl ssh console
app/bin/app remote
iex> App.Init.seeds()
```

Seeds added

**Comment by nelsonic** [comment]
Nice one @SimonLab 

Do we _need_ the "More content ... " noise under the selector: ?

image

**Comment by SimonLab** [comment]
We don't need it, however I've added it to show that the dropdown is displayed on top of the "more content" paragraph

**Comment by nelsonic** [comment]
What would be cool would be to display the selected people below the selector similar to how GitHub does it. 💭

**Pull request merged**